### ISSUE TITLE
CER and loss calculation for validation set during pure attention ASR (mtalpha=0)

### ISSUE THREAD

**Issue by vinitunni** [espnet/espnet#953]
I am trying ASR task with Mozilla common voice dataset using a pure attention model (mtalpha=0, ctc_weight=0). However, while calculating validation values, the cer and loss values are significantly higher (cer_valid~4.0 to 5.0). I manually calculater cer/wer using the printed groundtruth and predictions (when verbose=1) and these values are closer to 0.7.
To further check, I ran expts with mtalpha=0.01 and the cer values printed in this case are around the range of 0.7 (which seems plausible). Are there some extra settings we need to ensure when running a pure attention based task??

My network config is given below

```
2019-07-07 12:41:48,786 (deterministic_utils:24) INFO: torch type check is disabled
2019-07-07 12:41:48,888 (asr:234) INFO: #input dims : 249
2019-07-07 12:41:48,888 (asr:235) INFO: #output dims: 5015
2019-07-07 12:41:48,888 (asr:243) INFO: Pure attention mode
2019-07-07 12:41:48,888 (e2e_asr:85) WARNING: Subsampling is not performed for vgg*. It is performed in max pooling layers at CNN.
2019-07-07 12:41:48,888 (e2e_asr:86) INFO: subsample: 1 1 1 1 1 1
2019-07-07 12:41:50,395 (encoders:231) INFO: Use CNN-VGG + BLSTM for encoder
2019-07-07 12:41:52,677 (asr:267) INFO: writing a model config file to exp/train_us_pytorch_vggblstm_e5_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0_drop0.2_adadelta_sampprob0.0_bs12_mli600_mlo150_delta/results/model.json
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: accum_grad: 1
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: aconv_chans: 10
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: aconv_filts: 100
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: adim: 1024
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: aheads: 4
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: apply_uttmvn: True
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: atype: location
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: awin: 5
2019-07-07 12:41:52,743 (asr:271) INFO: ARGS: backend: pytorch
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: badim: 320
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: batch_bins: 0
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: batch_count: auto
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: batch_frames_in: 0
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: batch_frames_inout: 0
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: batch_frames_out: 0
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: batch_size: 12
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: bdropout_rate: 0.0
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: beam_size: 4
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: blayers: 2
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: bprojs: 300
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: btype: blstmp
2019-07-07 12:41:52,744 (asr:271) INFO: ARGS: bunits: 300


2019-07-07 12:41:52,746 (asr:271) INFO: ARGS: config: None
2019-07-07 12:41:52,746 (asr:271) INFO: ARGS: config2: None
2019-07-07 12:41:52,746 (asr:271) INFO: ARGS: config3: None
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: context_residual: False
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: criterion: acc
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: ctc_type: warpctc
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: ctc_weight: 0.3
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: debugdir: exp/train_us_pytorch_vggblstm_e5_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0_drop0.2_adadelta_sampprob0.0_bs12_mli600_mlo150_delta
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: debugmode: 1
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: dict: data/lang_char/train_us_unigram5000_units.txt
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: dlayers: 2
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: dropout_rate: 0.2
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: dropout_rate_decoder: 0.0
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: dtype: lstm
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: dunits: 1024
2019-07-07 12:41:52,747 (asr:271) INFO: ARGS: early_stop_criterion: validation/main/acc
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: elayers: 5
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: elayers_sd: 4
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: epochs: 20
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: eprojs: 1024
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: eps: 1e-08
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: eps_decay: 0.01
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: etype: vggblstm
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: eunits: 1024
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: fbank_fmax: None
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: fbank_fmin: 0.0
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: fbank_fs: 16000
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: grad_clip: 5
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: lm_weight: 0.1
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: lsm_type: 
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: lsm_weight: 0.0
2019-07-07 12:41:52,748 (asr:271) INFO: ARGS: maxlen_in: 600
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: maxlen_out: 150
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: maxlenratio: 0.0
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: minibatches: 0
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: minlenratio: 0.0
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: model_module: espnet.nets.pytorch_backend.e2e_asr:E2E
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: mtlalpha: 0.0
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: n_iter_processes: 0
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: n_mels: 80
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: nbest: 1
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: ngpu: 2
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: num_save_attention: 3
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: num_spkrs: 1
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: opt: adadelta
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: outdir: exp/train_us_pytorch_vggblstm_e5_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0_drop0.2_adadelta_sampprob0.0_bs12_mli600_mlo150_delta/results
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: patience: 12
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: penalty: 0.0
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: preprocess_conf: None
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: ref_channel: -1
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: report_cer: True
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: report_wer: True
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: resume: /home/vinit/exp/espnet-0.4.0/egs/MCVNew/asr1/exp/train_us_pytorch_vggblstm_e5_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0_drop0.2_adadelta_sampprob0.0_bs12_mli600_mlo150_delta/results/snapshot.ep.9
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: rnnlm: None
2019-07-07 12:41:52,749 (asr:271) INFO: ARGS: rnnlm_conf: None
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: sampling_probability: 0.0
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: seed: 1
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: sortagrad: 0
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: spa: False
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: stats_file: None
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: subsample: 1_2_2_1_1
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: sym_blank: <blank>
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: sym_space: <space>
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: tensorboard_dir: tensorboard/train_us_pytorch_vggblstm_e5_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0_drop0.2_adadelta_sampprob0.0_bs12_mli600_mlo150_delta
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: threshold: 0.0001
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: train_json: dump/train_us/deltatrue/data_unigram5000.json
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: use_beamformer: True
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: use_dnn_mask_for_wpe: False
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: use_frontend: False
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: use_wpe: False
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: uttmvn_norm_means: True
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: uttmvn_norm_vars: False
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: valid_json: dump/test_us/deltatrue/data_unigram5000.json
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: verbose: 1
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: wdropout_rate: 0.0
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: weight_decay: 0.0
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: wlayers: 2
2019-07-07 12:41:52,750 (asr:271) INFO: ARGS: wpe_delay: 3
2019-07-07 12:41:52,751 (asr:271) INFO: ARGS: wpe_taps: 5
2019-07-07 12:41:52,751 (asr:271) INFO: ARGS: wprojs: 300
2019-07-07 12:41:52,751 (asr:271) INFO: ARGS: wtype: blstmp
2019-07-07 12:41:52,751 (asr:271) INFO: ARGS: wunits: 300
2019-07-07 12:41:52,757 (asr:279) INFO: batch size is automatically increased (12 -> 24)
2019-07-07 12:41:57,734 (batchfy:320) INFO: count is auto detected as seq
2019-07-07 12:41:57,929 (batchfy:339) INFO: # utts: 68814
2019-07-07 12:41:57,940 (batchfy:376) INFO: # minibatches: 3115
2019-07-07 12:41:57,941 (batchfy:320) INFO: count is auto detected as seq
2019-07-07 12:41:57,949 (batchfy:339) INFO: # utts: 6237
2019-07-07 12:41:57,950 (batchfy:376) INFO: # minibatches: 278


```

**Comment by kan-bayashi** [comment]
> I manually calculater cer/wer using the printed groundtruth and predictions (when verbose=1) and these values are closer to 0.7.

Do you mean printed ones represent this part?
https://github.com/espnet/espnet/blob/9f0729f31bdf4ba6fdc5ca8a48042c9e6f24e00c/espnet/nets/pytorch_backend/rnn/decoders.py#L211-L212
Here, predictions are calculated using teacher-forcing, so it becomes lower than actual CER.

And in pure attention mode, it is better to set `minlenratio`, `maxlenratio`. `penalty` not to repeat the sentence or to avoid the too short sentence.

**Comment by vinitunni** [comment]
I did make the above changes. But looks like my main issue was that ctc_weight was not being passed during training args and the default value of 0.3 was being used. Thus higher values during validation/;
